dev_guide: Various small updates (#53273)
ansible/ansible: docs/docsite/rst/dev_guide/developing_modules_documenting.rst

==============

Every Ansible module must begin with ``#!/usr/bin/python`` - this "shebang" allows ``ansible_python_interpreter`` to work.

> This is immediatelly followed by ``# -*- coding: utf-8 -*-`` to ensure that it is clear that the file is UTF-8 encoded.

Reviewer: ```suggestion
This is immediately followed by ``# -*- coding: utf-8 -*-`` to clarify that the file is UTF-8 encoded.
```

.. _copyright:

Copyright and license
=====================

> After the shebang and coding, there should be a `copyright line <https://www.gnu.org/licenses/gpl-howto.en.html>`_ with the original copyright holder and a license declaration. The license declaration should be ONLY one line, not the full GPL prefix.:

Reviewer: ```suggestion
After the shebang and UTF-8 coding, there should be a `copyright line <https://www.gnu.org/licenses/gpl-howto.en.html>`_ with the original copyright holder and a license declaration. The license declaration should be ONLY one line, not the full GPL prefix.:
```

ANSIBLE_METADATA block
======================

> After the shebang, the coding, the copyright, and the license, your module file should contain an ``ANSIBLE_METADATA`` section. This section provides information about the module for use by other tools. For new modules, the following block can be simply added into your module:

Reviewer: ```suggestion
After the shebang, the UTF-8 coding, the copyright, and the license, your module file should contain an ``ANSIBLE_METADATA`` section. This section provides information about the module for use by other tools. For new modules, the following block can be simply added into your module:
```

DOCUMENTATION block
===================

> After the shebang, the coding, the copyright line, the license, and the ``ANSIBLE_METADATA`` section comes the ``DOCUMENTATION`` block. Ansible's online module documentation is generated from the ``DOCUMENTATION`` blocks in each module's source code. The ``DOCUMENTATION`` block must be valid YAML. You may find it easier to start writing your ``DOCUMENTATION`` string in an :ref:`editor with YAML syntax highlighting <other_tools_and_programs>` before you include it in your Python file. You can start by copying our `example documentation string <https://github.com/ansible/ansible/blob/devel/examples/DOCUMENTATION.yml>`_ into your module file and modifying it. If you run into syntax issues in your YAML, you can validate it on the `YAML Lint <http://www.yamllint.com/>`_ website.

Reviewer: ```suggestion
After the shebang, the UTF-8 coding, the copyright line, the license, and the ``ANSIBLE_METADATA`` section comes the ``DOCUMENTATION`` block. Ansible's online module documentation is generated from the ``DOCUMENTATION`` blocks in each module's source code. The ``DOCUMENTATION`` block must be valid YAML. You may find it easier to start writing your ``DOCUMENTATION`` string in an :ref:`editor with YAML syntax highlighting <other_tools_and_programs>` before you include it in your Python file. You can start by copying our `example documentation string <https://github.com/ansible/ansible/blob/devel/examples/DOCUMENTATION.yml>`_ into your module file and modifying it. If you run into syntax issues in your YAML, you can validate it on the `YAML Lint <http://www.yamllint.com/>`_ website.
```

  * A detailed description (generally two or more sentences).
  * Must be written in full sentences, i.e. with capital letters and periods/full stops.
  * Shouldn't mention the module name.
  * Make use of multiple entries rather than using one long paragraph.
  * Don't quote complete values unless it is required by YAML.

:version_added:

  * The version of Ansible when the module was added.

>   * This is a string, and not a float, i.e. ``version_added: '2.1'``

Reviewer 2: Doesn't seem necessary.

:author:

  * Name of the module author in the form ``First Last (@GitHubID)``.
  * Use a multi-line list if there is more than one author.
  * Don't use quotes as it should not be required by YAML.

:deprecated:

  * Marks modules that will be removed in future releases. See also :ref:`module_lifecycle`.

:options:

>   * Options are usually called `parameters` and sometimes arguments. We will be using `parameters` for the remainder of the documentation.

Reviewer: The documentation field is called `options`, not `parameters` - so it's a bit confusing to say "create a field called `options`" and then document it as if it were called `parameters`. I like the move toward consistency, for sure, but describing a field with one name using a different name seems like a step backward.

Author: Sure, but we use 'parameter' almost everywhere else (including the module documentation), so why complicate things in the documentation, just because it is called `options` in the module docs spec? I am all for renaming this part in Ansible too, but that won't happen in this PR.

Author: PS We are already having 3 names (option, argument, parameter) for the same thing, so I don't think my changes is *taking a step backward*. At least we are clarifying that it is the same thing, and we standardize on a single name. That is a step forward.

Reviewer 2: ansible ad-hoc uses `-a` to specify module args and there's `args:` for free-form modules...

  * If the module has no parameters (for example, it's a ``_facts`` module), all you need is one line: ``options: {}``.
  * If your module has parameters (in other words, accepts arguments), each parameter should be documented thoroughly. For each module parameter, include:

  :parameter-name:

    * Declarative operation (not CRUD), to focus on the final state, for example `online:`, rather than `is_online:`.
    * The name of the parameter should be consistent with the rest of the module, as well as other modules in the same category.

>     * In doubt, look for other modules to find parameter names that are used for the same purpose, we like to offer consistency to our users.

Reviewer 2: ```suggestion
    * When in doubt, look for other modules to find parameter names that are used for the same purpose; we like to offer consistency to our users.
```

  :description:

    * Detailed explanation of what this parameter does. It should be written in full sentences.

>     * The first entry is a description of the parameter itself, subsequent entries detail its use, possible value format or dependencies.

Reviewer 2: ```suggestion
    * The first entry is a description of the parameter itself; subsequent entries detail its use, possible value format or dependencies.
```

Documentation fragments
-----------------------

If you're writing multiple related modules, they may share common documentation, such as authentication details, file mode settings, ``notes:`` or ``seealso:`` entries. Rather than duplicate that information in each module's ``DOCUMENTATION`` block, you can save it once as a doc_fragment plugin and use it in each module's documentation. In Ansible, shared documentation fragments are contained in a ``ModuleDocFragment`` class in `lib/ansible/plugins/doc_fragments/ <https://github.com/ansible/ansible/tree/devel/lib/ansible/plugins/doc_fragments>`_. To include a documentation fragment, add ``extends_documentation_fragment: FRAGMENT_NAME`` in your module's documentation.

.. _note:

>   * in Ansible 2.8 the Ansible directories for doc fragments changed, see documentation of previous versions to find the old locations.

Reviewer 3: Can we just state that:
> prior to Ansible 2.8 this was in `lib/ansible/utils/module_docs_fragments`

Rather than making people have to look it up?

.. versionadded:: 2.8

> Since Ansible 2.8, you can have user supplied doc_fragments by using a ``doc_fragments`` directory adjacent to play or role, just like any other plugin.

Reviewer: ```suggestion
Since Ansible 2.8, you can have user-supplied doc_fragments by using a ``doc_fragments`` directory adjacent to play or role, just like any other plugin.
```

EXAMPLES block
==============

> After the shebang, the coding, the copyright line, the license, the ``ANSIBLE_METADATA`` section, and the ``DOCUMENTATION`` block comes the ``EXAMPLES`` block. Here you show users how your module works with real-world examples in multi-line plain-text YAML format. The best examples are ready for the user to copy and paste into a playbook. Review and update your examples with every change to your module.

Reviewer: ```suggestion
After the shebang, the UTF-8 coding, the copyright line, the license, the ``ANSIBLE_METADATA`` section, and the ``DOCUMENTATION`` block comes the ``EXAMPLES`` block. Here you show users how your module works with real-world examples in multi-line plain-text YAML format. The best examples are ready for the user to copy and paste into a playbook. Review and update your examples with every change to your module.
```

Per playbook best practices, each example should include a ``name:`` line::

    EXAMPLES = r'''
    - name: Ensure foo is installed
      modulename:
        name: foo
        state: present
    '''

The ``name:`` line should be capitalized and not include a trailing dot.

If your examples use boolean parameters, use yes/no values. Since the documentation generates boolean values as yes/no, having the examples use these values as well makes the module documentation more consistent.

If your module returns facts that are often needed, an example of how to use them can be helpful.

.. _return_block:

RETURN block
============

> After the shebang, the coding, the copyright line, the license, the ``ANSIBLE_METADATA`` section, ``DOCUMENTATION`` and ``EXAMPLES`` blocks comes the ``RETURN`` block. This section documents the information the module returns for use by other modules.

Reviewer: ```suggestion
After the shebang, the UTF-8 coding, the copyright line, the license, the ``ANSIBLE_METADATA`` section, ``DOCUMENTATION`` and ``EXAMPLES`` blocks comes the ``RETURN`` block. This section documents the information the module returns for use by other modules.
```

If your module doesn't return anything (apart from the standard returns), this section of your module should read: ``RETURN = r''' # '''``
Otherwise, for each value returned, provide the following fields. All fields are required unless specified otherwise.

:return name:
  Name of the returned field.

  :description:

>     Detailed description of what this value represents. Capitalized and with trailing dot.

Reviewer 2: Requiring capitalization and trailing dots only make sense (to me) when the description constitutes a complete sentence.

    One or more examples.
  :version_added:
    Only needed if this return was extended after initial Ansible release, i.e. this is greater than the top level `version_added` field.
    This is a string, and not a float, i.e. ``version_added: '2.3'``.
  :contains:
    Optional. To describe nested return values, set ``type: complex`` and repeat the elements above for each sub-field.

Here are two example ``RETURN`` sections, one with three simple fields and one with a complex nested field::

    RETURN = r'''
    dest:

>         description: Destination file/path

Reviewer: ```suggestion
        description: Destination file/path.
```

        returned: success
        type: str
        sample: /path/to/file.txt
    src:

>         description: Source file used for the copy on the target machine

Reviewer: ```suggestion
        description: Source file used for the copy on the target machine.
```

        returned: changed
        type: str
        sample: /home/httpd/.ansible/tmp/ansible-tmp-1423796390.97-147729857856000/source
    md5sum:

>         description: MD5 checksum of the file after running copy

Reviewer: ```suggestion
        description: MD5 checksum of the file after running copy.
```

Python imports
==============

> After the shebang, the coding, the copyright line, the license, and the sections for ``ANSIBLE_METADATA``, ``DOCUMENTATION``, ``EXAMPLES``, and ``RETURN``, you can finally add the python imports. All modules must use Python imports in the form:

Reviewer: ```suggestion
After the shebang, the UTF-8 coding, the copyright line, the license, and the sections for ``ANSIBLE_METADATA``, ``DOCUMENTATION``, ``EXAMPLES``, and ``RETURN``, you can finally add the python imports. All modules must use Python imports in the form:
```
